### ISSUE TITLE
Html::attr() and attr() skip attributes without value

### ISSUE THREAD

**Issue by trych** [k-next/kirby#4752]
## Description
If I try to use these helper methods for empty attributes like

```php
Html::attr(['autoplay' => '', 'controls' => ''])
```
the attributes will not be inserted at all.

**Expected behavior**  
The attributes should be inserted.
Ideally, even a syntax like
```php
Html::attr(['autoplay', 'controls'])
```
should be possible, but I imagine this could get complicated when non-value and value attributes are mixed. Either way, attributes with an empty value should *not* be skipped.

## Your setup
**Kirby 3.7.5**

**Comment by afbora** [comment]
I'm not sure which behavior would be more stable but as workaround you can set `true`;

````php
Html::attr(['autoplay' => true, 'controls' => true])
````

**Comment by trych** [comment]
Thanks for the quick response. I'd rather wait for a non-workaround solution though, so I don't need to change my projects later on again, in case the syntax will look different in the end. :)

**Comment by lukasbestle** [comment]
The current official solution is to use an empty space:

https://github.com/getkirby/kirby/blob/c7f52c4932d637c7eda090095285a73119a24456/src/Toolkit/Xml.php#L97-L99

So your code would look like this:

```php
Html::attr(['autoplay' => ' ', 'controls' => ' ']);
```

I'm also not happy about this. IMO an empty string should result in an existing but empty attribute, a space should result in a space, `null` should result in no attribute being present and `true` should result in `autoplay="autoplay"`. The last two are already implemented this way, the first two aren't.

However changing the supported argument values would be a pretty massive breaking change and it will be hard to migrate to this smoothly.

I absolutely love the idea of passing just values though. Those are easy to detect with `is_int($key)` (AFAIK HTML attributes must not start with a number).

**Comment by trych** [comment]
> However changing the supported argument values would be a pretty massive breaking change and it will be hard to migrate to this smoothly.

I might misunderstand you, but I doubt that any current setup would be broken by making
```php
Html::attr(['autoplay' => '', 'controls' => ''])
```
result in actually placing the passed attributes without value and thus behaving like one would expect.
I mean, since this does not work at all currently nobody should have a (non-functional) code like this in their project, so no production code will break imho.

**Comment by afbora** [comment]
The following example might be what Lukas means. If the developer has made a use like this, there will be a breaking change.

````php
Html::attr(['autoplay' => $autoplay ?? '']);
````

Don't think `Html::attr` just for video attributes, it could be used for any HTML attributes.

Btw I like the sending indexed array idea too. Maybe we can use the `A::isAssociative` method for this.

**Comment by lukasbestle** [comment]
Yes, the `Html` methods are also used throughout the core and may well be used like Ahmet wrote in production also in site or plugin code.

What we could do is a multi-step process:

1. First deprecate passing an empty string, instead suggest `null`.
2. Then switch over to an empty string meaning an empty attribute. At the same time deprecate passing a space, instead suggest an empty string.
3. Finally switch over a space meaning a space.

Each of these steps would need to happen in a major release. 

> Maybe we can use the `A::isAssociative` method for this.

I think we could even support mixed arrays and check each key separately. Then you could pass both attributes with values (array key/value) and without values (just the array value) at the same time.

**Comment by distantnative** [comment]
Since most of the current logic is happening in the `Xml` class and is only inherited by the `Html` class: Would all the discussed changes be valid/ok for XML as well? Or do we have to solve them separately in the `Html` class?

**Comment by lukasbestle** [comment]
> Would all the discussed changes be valid/ok for XML as well?

Yes, I think so. The weird "space means empty attribute" behavior feels like a hack/a method signature impurity to me, independent from the output format.

**Comment by distantnative** [comment]
How do we handle the new `Html::attr(['autoplay']);`. Currently, the class seems to want to render all as `autoplay=""`, but online I find rather that it should be just `autoplay` without any `=""`?

**Comment by distantnative** [comment]
Currently it seems to be:

| value | result |
|-|-|
| `' '` | `autoplay=""` |
| `true` | `autoplay="autoplay"` |
| `null`, `''`, `[]`, `false` | attribute won't be rendered |

**Comment by lukasbestle** [comment]
> How do we handle the new `Html::attr(['autoplay']);`

I think this one does depend on the output format. AFAIK, attributes without values are not valid in XML. So I'd say the output should be `autoplay` for HTML and `autoplay="autoplay"` for XML.

We can implement this by converting `int => 'autoplay'` to `'autoplay' => true`. Then the existing code will handle it for each format.

**Comment by distantnative** [comment]
Would this be a good first step then? https://github.com/getkirby/kirby/commit/8f56c09acb4f57ec9c9a2ddb8a5b4ea43d345a64

**Comment by lukasbestle** [comment]
Looks great. I have two minor suggestions:

- Deprecation wording:
  ```diff
  - In a future version, passing an empty string won\'t omit the attribute anymore but render it without value
  + In a future version, passing an empty string won\'t omit the attribute anymore but render it with an empty value. To omit the attribute, please pass `null`.
  ```
- Could you add TODO comments for 3.10 and 3.11 for the [next steps](https://github.com/getkirby/kirby/issues/4752#issuecomment-1267969344) to keep track of them so they don't get lost?

**Comment by distantnative** [comment]
✅ https://github.com/getkirby/kirby/pull/4935 is merged for 3.8.4, https://github.com/getkirby/kirby/pull/4934 is ready for 3.9 (and with to-dos for 3.10, 3.11)
